src/controllers/trainer.js

```diff
@@ -22,4 +22,20 @@ export const trainerLoginController = async (loginDetails) => {
         return response;
     }
     return false;
-} 
\ No newline at end of file
+}
+
+export const listLeaveController = (id, date, FN_or_AN) => {
+    return listLeaveModel(id, date, FN_or_AN)
+}
+
+export const updateLeaveController = (id, date, FN_or_AN) => {
+    return updateLeaveModel(id, date, FN_or_AN)
+}
+
+export const insertLeaveController = (date, FN_or_AN) => {
+    return insertLeaveModel(date, FN_or_AN)
```

Comment:
For insert and update, put a check if the trainer has already marked the current session as leave, and only proceed if there is not one already. Write a function that does this check and returns a boolean value and call this function in both edit and insert controllers.